File: objective-c/console/ttt_engine/Src/Game.m
#import "Game.h"
#import "BoardLogic.h"
#import "Players/PlayerFactory.h"

@implementation Game

- (id) initWithUi:(id)ui {
    if ((self = [super init])) {
        self.ui = ui;
    }
    return (self);
}

- (void) createBoard {
    int boardSize = [[[self ui] promptBoardSize] intValue];
    self.board = [[Board alloc] init:boardSize];
}

- (void) createPlayerOne {
    NSString *playerType = [[self ui] promptPlayerOneType];
    self.playerOne = [PlayerFactory create:playerType withMark:@"X"];
}

- (void) createPlayerTwo {
    NSString *playerType = [[self ui] promptPlayerTwoType];
    self.playerTwo = [PlayerFactory create:playerType withMark:@"O"];
}

- (void) play {
    [self startGame];

    while (![BoardLogic isOver:self.board]) {
        [self playGame];
    }

    [self endGame];
}

- (void) startGame {
    [[self ui] welcomeUser];

    [self setUpGame];
}

- (void) setUpGame {
    [self createBoard];
    [self createPlayerOne];
    [self createPlayerTwo];
}

- (void) playGame {
    [[self ui] printBoard:self.board];
    [self makePlayerMove:self.playerOne];

    if (![BoardLogic isOver:self.board]) {
        [[self ui] printBoard:self.board];
        [self makePlayerMove:self.playerTwo];
    }
}

- (void) makePlayerMove:(id<Player>)player {
    self.board.slots = [player makeMoveWith:self.board andUi:self.ui];
}

- (void) endGame {
    [[self ui] printBoard:self.board];
    [[self ui] promptResult:[BoardLogic gameState:self.board]];

    if ([[self ui] promptPlayAgain]) {
        [self play];
    }
    [[self ui] promptGoodBye];
}

@end

File: javascript/engine/src/Game.js
function Game(ui) {
  this.ui = ui;

  this.start = function() {
    this.ui.welcomePlayer();

    this.createBoard();
    this.createPlayerOne();
    this.createPlayerTwo();
  }

  this.end = function() {
    this.ui.farewellPlayer();
  }

  this.createBoard = function() {
    boardSize = this.ui.askBoardSize();
    this.board = new Board(boardSize);
  }

  this.createPlayerOne = function() {
    playerType = this.ui.askPlayerOneType();
    this.playerOne = new Player("X", playerType, ui);
  }

  this.createPlayerTwo = function() {
    playerType = this.ui.askPlayerTwoType();
    this.playerTwo = new Player("O", playerType, ui);
  }
}
